storage/pruner/pruner.go

```diff
@@ -0,0 +1,374 @@
+// base height for pruning.
+func (p *Pruner) Register() *BlockProcessedProducer {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	// Note: the initial height is 0 so pruning will not start until all producers have completed
+	// processing at least one block
+	producer := NewBlockProcessedProducer()
+	p.producers = append(p.producers, producer)
+
+	return producer
+}
+
+// baseHeight returns the highest height that all block processors have completed processing
+// for instance, if consensus is at 1000, and execution is at 750, the base height should be 750
+// this ensures all subsystems have the data they need regardless of the pruning thresholds used
+func (p *Pruner) baseHeight() uint64 {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	lowest := uint64(0)
+	for _, producer := range p.producers {
+		height := producer.highestCompleteHeight.Load()
+		if height < lowest {
+			lowest = height
+		}
+	}
+
+	return lowest
+}
+
+// initProgress initializes the pruned height consumer progress
+// If the progress is not found, it is initialized to the root sealed block height
+// No errors are expected during normal operation.
+func (p *Pruner) initProgress() (uint64, error) {
+	lowestHeight, err := p.progress.ProcessedIndex()
+	if err == nil {
+		return lowestHeight, nil
+	}
+
+	if !errors.Is(err, storage.ErrNotFound) {
+		return 0, fmt.Errorf("could not get processed index: %w", err)
+	}
+
+	err = p.progress.InitProcessedIndex(p.rootSealedHeader.Height)
+	if err != nil && !errors.Is(err, storage.ErrAlreadyExists) {
+		return 0, fmt.Errorf("could not init processed index: %w", err)
+	}
+
+	p.log.Warn().
+		Uint64("processed index", lowestHeight).
+		Msg("processed index not found, initialized.")
+
+	return p.progress.ProcessedIndex()
+}
+
+func (p *Pruner) loop(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	lowestHeight, err := p.initProgress()
+	if err != nil {
+		ctx.Throw(err)
+		return
+	}
+
+	p.lowestHeader, err = p.headers.ByHeight(lowestHeight)
+	if err != nil {
+		ctx.Throw(fmt.Errorf("could not get header for lowest height (%d): %w", lowestHeight, err))
+		return
+	}
+
+	ready()
+
+	ticker := time.NewTicker(DefaultPruneInterval)
+	defer ticker.Stop()
+
+	for ctx.Err() == nil {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+		}
+
+		// baseHeight is the highest height for which we have completed all processing
+		baseHeight := p.baseHeight()
+		if baseHeight < p.lowestHeader.Height {
+			// this should never happen and means there's a bug. the lowest fully processed height
+			// cannot be lower than the lowest height in the database
+			ctx.Throw(errors.New("inconsistent state: base height must not be lower than the lowest height"))
+			return
+		}
+
+		lg := p.log.With().Uint64("baseHeight", baseHeight).Logger()
+
+		// if there are not enough blocks to prune, skip
+		if baseHeight-p.lowestHeader.Height < p.pruneThreshold {
+			lg.Debug().
```

Comment:
I think the pruning interval is usually large enough that we can afford an info log, which makes it easy to debug as its acting as a heartbeat. We could include as many information as possible.